index.d.ts

```diff
@@ -0,0 +1,88 @@
+// Type definitions for express-session 1.16
+// Project: https://github.com/expressjs/session
+// Definitions by: Hiroki Horiuchi <https://github.com/horiuchi>
+//                 Jacob Bogers <https://github.com/jacobbogers>
+//                 Naoto Yokoyama <https://github.com/builtinnya>
+//                 Ryan Cannon <https://github.com/ry7n>
+//                 HoldYourWaffle <https://github.com/HoldYourWaffle>
+
+import express from 'express'
+import node from 'events'
+
+
+
+declare global {
+  namespace Express {
+    interface Request { //inject additional properties on express.Request
+      session?: Session;
+      sessionID?: string;
+    }
+  }
+
+  export interface SessionData {
+    //seperate interface to allow application-specific declaration merging (see README)
+  }
+}
+
+
+
+interface SessionOptions {
+  secret: string | string[];
+
+  genid?(req: express.Request): string;
+  name?: string;
+  store?: Store;
+  cookie?: express.CookieOptions;
+  rolling?: boolean;
+  resave?: boolean;
+  proxy?: boolean;
+  saveUninitialized?: boolean;
+  unset?: string;
+}
+
+export default function session(options?: SessionOptions): express.RequestHandler;
+
+
+
+interface Session extends SessionData {
+  id: string;
+  cookie: SessionCookie;
+
+  regenerate(callback: (err: any) => void): Session;
+  destroy(callback: (err: any) => void): Session;
+  reload(callback: (err: any) => void): Session;
+  save(callback?: (err: any) => void): Session;
+  touch(): Session;
+}
+
+interface SessionCookie extends express.CookieOptions { //XXX actually a class
+  originalMaxAge: number;
+}
+
+
+
+export abstract class Store extends node.EventEmitter {
+  constructor(config?: any);
+
+  regenerate: (req: express.Request, callback: (err?: any) => any) => void;
+  load: (sid: string, callback: (err: any, session?: SessionData) => any) => void;
+  createSession: (req: express.Request, sess: SessionData) => void;
+
+  abstract get: (sid: string, callback: (err: any, session?: SessionData | null) => void) => void;
+  abstract set: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
+  abstract destroy: (sid: string, callback?: (err?: any) => void) => void;
+  abstract all: (callback: (err: any, obj?: { [sid: string]: SessionData; } | null) => void) => void;
```

Comment:
So forgive me If I don't understand and this may be a non-issue: there are methods declared here that stores are not required to implement (https://github.com/expressjs/session#session-store-implementation). Would this cause issues if extended and they are not implemented?

Reply:
Good catch. I made `all`, `length`, `clear` and `touch` optional.